### PR TITLE
Replace `env_logger` with `tracing`

### PR DESCRIPTION
**Pull request by ffuugoo**
This PR replaces `env_logger` with `tracing`/`tracing-subscriber`/`tracing-log` as our `log` backend (and also removes half of the custom features added in #2187).

(Currently based on top of #2187. Would be rebased onto `dev` once #2187 is merged.)

__TODO:__
- [x] Fix `DEVELOPMENT.md` documentation after removing custom tracing-related features from `qdrant` crate
- [x] ~~Make `tracing` _required_ dependency of `lib/*` crates and remove custom `tracing` feature everywhere?~~
  - There's practically no downside to leaving `tracing` optional in `lib/*` crates, so I crossed it out.

__Further Improvements:__
- Add API to change log-level/filters when Qdrant is already running (#2383)
  - should be pretty trivial to do with [`tracing_subscriber::reload`]
- Add configuration parameter to [select "span events" to be printed][with-span-events]
  - (and an API to change this in runtime)
  - span events are rather verbose to enable them by default, but they are _extremely_ useful for debugging
  - adding configuration parameter should be trivial
  - adding an API - _moderately_ annoying
- Add file appender (and all required configuration/infrastructure), so that logs can be written to a file (#2392)
  - [`RollingFileAppender`] from [`tracing-appender`] crate does not _yet_ support _practical_ rollover strategies
    - https://github.com/tokio-rs/tracing/pull/2323 is merged, but not yet released
    - https://github.com/tokio-rs/tracing/pull/2497 is not yet merged
  - there's also [`tracing-rolling-file`], which looks fine, but I haven't tested it yet
- Implement slog drain that would convert slog records into tracing events, _preserving slog's "structured" KV pairs_ 
  - [`slog-stdlog`] (that we are currently using) is fine
    - but it does not preserve KV pairs
    - I'm not sure if `raft` crate actually emits any, but it would be kinda silly to use `slog` if it does not, so I assume it does
    - and so I'd want to capture/log these KV pairs
  - [`tracing-slog`] is mostly useless
    - it is extremely simplistic and does not preserve KV pairs
    - which is, like, the whole point of `slog` and any _specific_ "slog-to-tracing" bridge
    - because [`slog-stdlog`] is perfectly fine if you don't want/need KV pairs
- Add configuration parameter to select output format?
  - (and an API to change this in runtime)
  - I can already tell, that this will be __super__ annoying to implement, so not sure if it's worth it

[`tracing_subscriber::reload`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/reload/index.html
[with-span-events]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#method.with_span_events
[`RollingFileAppender`]: https://docs.rs/tracing-appender/latest/tracing_appender/rolling/struct.RollingFileAppender.html
[`tracing-appender`]: https://docs.rs/tracing-appender/latest/tracing_appender/index.html
[`tracing-rolling-file`]: https://docs.rs/tracing-rolling-file/latest/tracing_rolling_file/
[`slog-stdlog`]: https://docs.rs/slog-stdlog/latest/slog_stdlog/
[`tracing-slog`]: https://docs.rs/tracing-slog/latest/tracing_slog/

### All Submissions:

* [ ] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: src/tracing.rs [hunk 1]
```diff
@@ -1,55 +1,40 @@
...
+    );
+
+    // Use `console` or `console-subscriber` feature to enable `console-subscriber`
+    //
+    // Note, that `console-subscriber` requires manually enabling
+    // `--cfg tokio_unstable` rust flags during compilation!
+    //
+    // Otherwise `console_subscriber::spawn` call panics!
+    //
+    // See https://docs.rs/tokio/latest/tokio/#unstable-features
```
Comment (timvisee) [inline src/tracing.rs:21]: Could you elaborate a bit on what this means. Is this only relevant for interacting with Tokio Console?

**Review by timvisee — APPROVED** [review]
Nice work!

In development mode it is a lot more noisy now. Can we disable logging on some modules by default to reduce it a bit? Or do you think keeping these is more practical?

Before:

image

After:

image

**Comment by ffuugoo** [comment]
> In development mode it is a lot more noisy now. Can we disable logging on some modules by default to reduce it a bit? Or do you think keeping these is more practical?

Some crates are using `tracing` instead of `log`, so we are _suddenly_ seeing these logs now. I agree that it makes sense to disable some of the more verbose logs. I personally use `debug,hyper=info,h2=info,tower=info,rustls=info`.

**Comment by timvisee** [comment]
> I personally use `debug,hyper=info,h2=info,tower=info,rustls=info`.

Yeah, that seems more reasonable.

Maybe we can set that as default [here](https://github.com/qdrant/qdrant/blob/47a03091f1e096ef7a34f481b5e3bf56522c612c/config/development.yaml#L2). What do you think?

File: src/tracing.rs (hunk 1 above)
Comment (ffuugoo) [reply src/tracing.rs:21]: > Is this only relevant for interacting with Tokio Console?

TL;DR: Yes.

There's a `tracing` feature in Tokio, but it's considered _experimental_, so if you want to enable it, you have to also explicitly set `RUSTFLAGS='--cfg tokio_unstable'`. And, this feature/flag combo simply enables `tracing` in Tokio. Like, tracing-spans/additional logs/etc.

It's _required_ for Tokio Console, but it _might_ also be useful for general Tokio debugging sometime. If you only want Tokio `tracing`, but without Tokio Console, you can enable `tokio-tracing` feature of the `qdrant` crate, but you will still have to set `RUSTFLAGS='--cfg tokio_unstable'`.

**Review by generall — APPROVED** [review]

**Review by generall — APPROVED** [review]
